File: core/cache.py
import hashlib
from pathlib import Path


class CorrectionCache:
    def __init__(self, cache_dir: str = ".cache/correzioni"):
        self.cache_dir = Path(cache_dir)
        self.cache_dir.mkdir(parents=True, exist_ok=True)
        self._mem_cache: dict[str, str] = {}

    def _hash(self, text: str) -> str:
        return hashlib.md5(text.encode("utf-8")).hexdigest()

    def get(self, text: str) -> str | None:
        if text in self._mem_cache:
            return self._mem_cache[text]
        h = self._hash(text)
        p = self.cache_dir / f"{h}.txt"
        if p.exists():
            val = p.read_text("utf-8").strip()
            self._mem_cache[text] = val
            return val
        return None

    def set(self, original: str, corrected: str):
        if original == corrected:
            return
        h = self._hash(original)
        p = self.cache_dir / f"{h}.txt"
        p.write_text(corrected.strip(), "utf-8")
        self._mem_cache[original] = corrected.strip()

    def get_or_correct(self, text: str, correct_fn) -> str:
        cached = self.get(text)
        if cached is not None:
            return cached
        corrected = correct_fn(text)
        self.set(text, corrected)
        return corrected

File: crates/videoforge-core/src/cache.rs
use std::collections::HashMap;
use std::path::PathBuf;
use std::sync::RwLock;

pub struct CorrectionCache {
    cache_dir: PathBuf,
    mem_cache: RwLock<HashMap<String, String>>,
}

impl CorrectionCache {
    pub fn new(cache_dir: Option<PathBuf>) -> Self {
        let dir = cache_dir.unwrap_or_else(|| {
            let home = dirs::home_dir().unwrap_or_default();
            home.join(".cache/correzioni")
        });
        std::fs::create_dir_all(&dir).ok();
        Self {
            cache_dir: dir,
            mem_cache: RwLock::new(HashMap::new()),
        }
    }

    fn hash(&self, text: &str) -> String {
        let hash = blake3::hash(text.as_bytes());
        hash.to_hex()[..32].to_string()
    }

    pub fn get(&self, text: &str) -> Option<String> {
        if let Some(cached) = self.mem_cache.read().ok()?.get(text).cloned() {
            return Some(cached);
        }
        let h = self.hash(text);
        let path = self.cache_dir.join(format!("{h}.txt"));
        if let Ok(val) = std::fs::read_to_string(&path) {
            let trimmed = val.trim().to_string();
            if let Ok(mut mem) = self.mem_cache.write() {
                mem.insert(text.to_string(), trimmed.clone());
            }
            Some(trimmed)
        } else {
            None
        }
    }

    pub fn set(&self, original: &str, corrected: &str) {
        if original == corrected {
            return;
        }
        if let Ok(mut mem) = self.mem_cache.write() {
            mem.insert(original.to_string(), corrected.to_string());
        }
        let h = self.hash(original);
        let path = self.cache_dir.join(format!("{h}.txt"));
        std::fs::write(&path, corrected.trim()).ok();
    }

    pub fn get_or_correct(&self, text: &str, correct_fn: impl Fn(&str) -> String) -> String {
        if let Some(cached) = self.get(text) {
            return cached;
        }
        let corrected = correct_fn(text);
        self.set(text, &corrected);
        corrected
    }

    pub fn clear(&self) {
        if let Ok(mut mem) = self.mem_cache.write() {
            mem.clear();
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cache_store_and_retrieve() {
        let cache = CorrectionCache::new(None);
        cache.set("hello", "world");
        assert_eq!(cache.get("hello"), Some("world".to_string()));
    }

    #[test]
    fn test_cache_returns_nil_for_missing() {
        let cache = CorrectionCache::new(None);
        assert_eq!(cache.get("missing"), None);
    }

    #[test]
    fn test_cache_skips_identical() {
        let cache = CorrectionCache::new(None);
        cache.set("same", "same");
        assert_eq!(cache.get("same"), None);
    }

    #[test]
    fn test_cache_get_or_correct() {
        let cache = CorrectionCache::new(None);
        let result = cache.get_or_correct("test", |s| s.to_uppercase());
        assert_eq!(result, "TEST");
    }

    #[test]
    fn test_cache_persistence() {
        let dir = std::env::temp_dir().join("test_cache_persist");
        std::fs::create_dir_all(&dir).ok();
        {
            let cache = CorrectionCache::new(Some(dir.clone()));
            cache.set("ciao", "hello");
        }
        let cache2 = CorrectionCache::new(Some(dir.clone()));
        assert_eq!(cache2.get("ciao"), Some("hello".to_string()));
        std::fs::remove_dir_all(&dir).ok();
    }
}
